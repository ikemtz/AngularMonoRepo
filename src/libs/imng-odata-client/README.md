# imng-odata-client

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `nx test imng-odata-client` to execute the unit tests.
