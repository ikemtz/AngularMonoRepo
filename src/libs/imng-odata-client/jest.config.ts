import type { Config } from 'jest';

const config: Config = {
  displayName: 'imng-odata-client',
  reporters: [
    'default',
    [
      'jest-junit',
      {
        outputDirectory: '.',
        outputName: './junit.imng-odata-client.xml',
      },
    ],
  ],
  setupFilesAfterEnv: ['<rootDir>/src/test-setup.ts'],
  globals: {},
  coverageDirectory: '../../coverage/libs/imng-odata-client',
  transform: {
    '^.+\\.(ts|mjs|js|html)$': [
      'jest-preset-angular',
      {
        tsconfig: '<rootDir>/tsconfig.spec.json',
        stringifyContentPathRegex: '\\.(html|svg)$',
      },
    ],
  },
  transformIgnorePatterns: ['node_modules/(?!.*\\.mjs$)'],
  snapshotSerializers: [
    'jest-preset-angular/build/serializers/no-ng-attributes',
    'jest-preset-angular/build/serializers/ng-snapshot',
    'jest-preset-angular/build/serializers/html-comment',
  ],
  preset: '../../jest.preset.js',
};

export default config;
