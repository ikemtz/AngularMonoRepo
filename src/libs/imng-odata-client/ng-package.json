{
  "$schema": "../../node_modules/ng-packagr/ng-package.schema.json",
  "dest": "../../dist/libs/imng-odata-client",
  "lib": {
    "entryFile": "src/index.ts"
  }
}
