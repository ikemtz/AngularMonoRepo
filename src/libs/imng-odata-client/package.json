{
  "name": "imng-odata-client",
  "version": "0.0.1",
  "peerDependencies": {
    "@angular/common": ">= 14",
    "@angular/core": ">= 14",
    "imng-nrsrx-client-utils": "*",
    "rxjs": "*"
  },
  "dependencies": {
    "tslib": "^2.3.0"
  },
  "keywords": [
    "Angular",
    "OData"
  ],
  "license": "MIT",
  "repository": {
    "type": "git",
    "url": "https://github.com/ikemtz/AngularMonoRepo.git"
  },
  "icon": "https://avatars.githubusercontent.com/u/6444182?v=4",
  "private": false,
  "author": {
    "name": "Isaac Martinez <@ikemtz>",
    "url": "https://github.com/ikemtz"
  }
}
