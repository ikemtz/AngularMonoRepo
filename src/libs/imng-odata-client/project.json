{
  "projectType": "library",
  "sourceRoot": "libs/imng-odata-client/src",
  "prefix": "imng",
  "targets": {
    "build": {
      "executor": "@nrwl/angular:package",
      "outputs": ["dist/libs/imng-odata-client"],
      "options": {
        "project": "libs/imng-odata-client/ng-package.json"
      },
      "configurations": {
        "production": {
          "tsConfig": "libs/imng-odata-client/tsconfig.lib.prod.json"
        },
        "development": {
          "tsConfig": "libs/imng-odata-client/tsconfig.lib.json"
        }
      },
      "defaultConfiguration": "production"
    },
    "test": {
      "executor": "@nrwl/jest:jest",
      "outputs": ["coverage/libs/imng-odata-client"],
      "options": {
        "jestConfig": "libs/imng-odata-client/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "lint": {
      "executor": "@nrwl/linter:eslint",
      "options": {
        "lintFilePatterns": [
          "libs/imng-odata-client/**/*.ts",
          "libs/imng-odata-client/**/*.html"
        ]
      }
    }
  },
  "tags": []
}
