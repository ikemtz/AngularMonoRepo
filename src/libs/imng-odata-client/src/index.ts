export * from './lib/imng-odata-client.module';
export * from './lib/models';
export * from './lib/services/odata-client.service';
export * from './lib/helpers/get-filter-operator';
export * from './lib/helpers/create-odata-result';
