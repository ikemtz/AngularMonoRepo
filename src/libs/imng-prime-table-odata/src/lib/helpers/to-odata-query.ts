import {
  CompositeFilter,
  Filter,
  getFilterOperator,
  ODataQuery,
  Sort,
} from 'imng-odata-client';
import { PrimeTableState } from '../models/prime-odata-table-state';

export function toODataQuery(val: PrimeTableState): ODataQuery {
  const query: ODataQuery = {
    ...val,
    orderBy: val.multiSortMeta?.map(
      (x): Sort => ({
        field: x.field,
        dir: x.order === -1 ? 'desc' : 'asc',
      }),
    ),
    skip: val.first,
    top: val.rows,
  };

  if (val.filters) {
    const filters = Object.keys(val.filters || {})
      .map((key) => ({
        key,
        collection: val.filters?.[key] || [],
      }))
      .filter((t) => t.collection?.filter((t) => t.value).length)
      .map(
        (t): CompositeFilter => ({
          logic: t.collection?.every((e) => e.operator === 'and')
            ? 'and'
            : 'or',
          filters: t.collection.map(
            (m): Filter => ({
              field: t.key,
              operator: getFilterOperator(m.matchMode || 'eq'),
              value: m.value,
            }),
          ),
        }),
      );
    if (filters?.length) {
      query.filter = { logic: 'and', filters };
    }
  }
  return query;
}
