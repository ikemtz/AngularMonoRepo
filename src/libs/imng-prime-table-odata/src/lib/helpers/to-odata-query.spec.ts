import { ODataClientService } from 'imng-odata-client';
import { toODataQuery } from './to-odata-query';
import { HttpClient } from '@angular/common/http';
import { of } from 'rxjs';
import { TestBed } from '@angular/core/testing';

describe('toODataQuery', () => {
  it('should convert empty State', () => {
    const result = toODataQuery({});
    expect(result).toMatchSnapshot();
  });
  it('should convert typical State', () => {
    const result = toODataQuery({
      multiSortMeta: [
        { field: 'abc', order: 1 },
        { field: 'xyz', order: -1 },
      ],
      filters: {
        abc: [{ operator: 'all', matchMode: 'equals', value: 'xyz' }],
      },
    });
    expect(result).toMatchSnapshot();
  });
  it('should convert state with Child Filter', () => {
    const result = toODataQuery({
      multiSortMeta: [
        { field: 'abc', order: 1 },
        { field: 'xyz', order: -1 },
      ],
      filters: {
        ['abc.bdf/xyz']: [
          { operator: 'all', matchMode: 'equals', value: '123' },
        ],
      },
    });
    expect(result).toMatchSnapshot();
  });
  it('should convert state with parent Filters', () => {
    TestBed.configureTestingModule({
      providers: [
        ODataClientService,
        {
          provide: HttpClient,
          useValue: { get: jest.fn(() => of({})) },
        },
      ],
    }).compileComponents();
    const svc = TestBed.inject(ODataClientService);
    const query = toODataQuery({
      multiSortMeta: [
        { field: 'abc', order: 1 },
        { field: 'xyz', order: -1 },
      ],
      filters: {
        ['abc.bdf.xyz']: [
          { operator: 'all', matchMode: 'equals', value: '123' },
        ],
      },
    });
    const odataString = svc.getODataString(query);
    expect(odataString).toBe(
      `$filter=((abc/bdf/xyz+eq+'123'))&$orderby=abc,xyz+desc&$count=true`,
    );
  });
  it('should convert with no filters', () => {
    const result = toODataQuery({
      multiSortMeta: [
        { field: 'abc', order: 1 },
        { field: 'xyz', order: -1 },
      ],
    });
    expect(result).toMatchSnapshot();
  });
  it('should convert no sorting', () => {
    const result = toODataQuery({
      filters: {
        abc: [{ operator: 'all', matchMode: 'equals', value: 'xyz' }],
      },
    });
    expect(result).toMatchSnapshot();
  });
});
