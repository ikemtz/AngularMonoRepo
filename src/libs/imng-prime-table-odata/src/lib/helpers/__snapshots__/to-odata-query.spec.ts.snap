// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toODataQuery should convert empty State 1`] = `
Object {
  "filter": undefined,
  "orderBy": undefined,
  "skip": undefined,
  "top": undefined,
}
`;

exports[`toODataQuery should convert no sorting 1`] = `
Object {
  "filter": Object {
    "filters": Array [
      Object {
        "filters": Array [
          Object {
            "field": "abc",
            "operator": Object {
              "name": "equals",
              "toODataString": [Function],
            },
            "value": "xyz",
          },
        ],
        "logic": "or",
      },
    ],
    "logic": "and",
  },
  "orderBy": undefined,
  "skip": undefined,
  "top": undefined,
}
`;

exports[`toODataQuery should convert state with Child Filter 1`] = `
Object {
  "filter": Object {
    "filters": Array [
      Object {
        "filters": Array [
          Object {
            "childTable": "abc/bdf",
            "field": "xyz",
            "linqOperation": "any",
            "operator": Object {
              "name": "equals",
              "toODataString": [Function],
            },
            "value": "123",
          },
        ],
        "logic": "or",
      },
    ],
    "logic": "and",
  },
  "orderBy": Array [
    Object {
      "dir": "asc",
      "field": "abc",
    },
    Object {
      "dir": "desc",
      "field": "xyz",
    },
  ],
  "skip": undefined,
  "top": undefined,
}
`;

exports[`toODataQuery should convert typical State 1`] = `
Object {
  "filter": Object {
    "filters": Array [
      Object {
        "filters": Array [
          Object {
            "field": "abc",
            "operator": Object {
              "name": "equals",
              "toODataString": [Function],
            },
            "value": "xyz",
          },
        ],
        "logic": "or",
      },
    ],
    "logic": "and",
  },
  "orderBy": Array [
    Object {
      "dir": "asc",
      "field": "abc",
    },
    Object {
      "dir": "desc",
      "field": "xyz",
    },
  ],
  "skip": undefined,
  "top": undefined,
}
`;

exports[`toODataQuery should convert with no filters 1`] = `
Object {
  "filter": undefined,
  "orderBy": Array [
    Object {
      "dir": "asc",
      "field": "abc",
    },
    Object {
      "dir": "desc",
      "field": "xyz",
    },
  ],
  "skip": undefined,
  "top": undefined,
}
`;
